src/misc/messages.cpp

```diff
@@ -21,227 +21,751 @@
 
 #include "dosbox.h"
 
-#include <clocale>
-#include <cstdio>
-#include <cstdlib>
-#include <cstring>
-#include <map>
-#include <vector>
-
-#include "std_filesystem.h"
-#include <string>
-#include <unordered_map>
-
+#include "../dos/dos_locale.h"
 #include "ansi_code_markup.h"
 #include "checks.h"
 #include "control.h"
 #include "cross.h"
 #include "fs_utils.h"
 #include "host_locale.h"
 #include "setup.h"
+#include "std_filesystem.h"
 #include "string_utils.h"
 #include "support.h"
 #include "unicode.h"
 
-#define LINE_IN_MAXLEN 2048
+#include <cctype>
+#include <fstream>
+#include <map>
+#include <set>
+#include <string>
 
 CHECK_NARROWING();
 
-static const char *msg_not_found = "Message not Found!\n";
+static const std::string MsgNotFound = " MESSAGE NOT FOUND! ";
+static const std::string MsgNotValid = " MESSAGE NOT VALID! ";
+
+// ***************************************************************************
+// Single message storage class
+// ***************************************************************************
 
 class Message {
-private:
-	std::string markup_msg = {};
-	std::string rendered_msg = {};
+public:
+	// Note: any message needs to be verified before it can be safely used!
 
-	std::map<uint16_t, std::string> rendered_msg_by_codepage = {};
+	// Constructor for original, English strings
+	Message(const MsgFlags flags, const std::string& text);
+	// Constructor for translated strings from external file
+	Message(const std::string& text);
 
-	static std::string to_dos(const std::string& in_str, const uint16_t code_page)
+	const std::string& Get(const bool raw_requested);
+
+	bool IsValid() const
 	{
-		return utf8_to_dos(in_str,
-		                   DosStringConvertMode::WithControlCodes,
-		                   UnicodeFallback::Box,
-		                   code_page);
+		return is_verified && is_ok;
 	}
-
-	const char* CachedRenderString(const std::string& msg,
-	                               std::map<uint16_t, std::string>& output_msg_by_codepage)
+	void MarkInvalid()
 	{
-		assert(msg.length());
-		const uint16_t cp = get_utf8_code_page();
-		if (output_msg_by_codepage[cp].empty()) {
-			output_msg_by_codepage[cp] = to_dos(msg, cp);
-			assert(output_msg_by_codepage[cp].length());
+		is_ok = false;
+	}
+
+	// Use this one for English messages only
+	void Verify(const std::string& name);
+	// Use this one for translated messages
+	void Verify(const std::string& name, const Message& english);
+
+private:
+	Message() = delete;
+
+	std::string GetLogStart(const std::string& name) const;
+
+	void VerifyMessage(const std::string& name);
+	void VerifyFormatString(const std::string& name);
+	void VerifyFormatStringAgainst(const std::string& name,
```

Comment:
Should `English` rather be part of the function's name?

Reply:
Not here - `VerifyMessage` and `VerifyFormatString` are common for both English and translated messages.